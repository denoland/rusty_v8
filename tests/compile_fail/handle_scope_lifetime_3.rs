// Copyright 2019-2020 the Deno authors. All rights reserved. MIT license.
use std::pin::pin;

pub fn main() {
  let mut isolate = v8::Isolate::new(mock());
  let scope1 = pin!(v8::HandleScope::new(&mut isolate));
  let mut scope1 = scope1.init();
  let context = v8::Context::new(&mut scope1, v8::ContextOptions::default());
  let mut context_scope = v8::ContextScope::new(&mut scope1, context);

  let _local = {
    let mut _scope2 =
      v8::EscapableHandleScope::new((&mut *context_scope).as_mut());
    v8::Integer::new(&mut scope1, 123)
  };
}

fn mock<T>() -> T {
  unimplemented!()
}
