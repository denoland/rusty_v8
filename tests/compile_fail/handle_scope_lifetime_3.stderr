error[E0499]: cannot borrow `scope1` as mutable more than once at a time
  --> tests/compile_fail/handle_scope_lifetime_3.rs:14:22
   |
9  |   let mut context_scope = v8::ContextScope::new(&mut scope1, context);
   |                                                 ----------- first mutable borrow occurs here
...
14 |     v8::Integer::new(&mut scope1, 123)
   |                      ^^^^^^^^^^^ second mutable borrow occurs here
15 |   };
   |   - first borrow might be used here, when `_scope2` is dropped and runs the `Drop` code for type `v8::scope2::ScopeStorage`
